File: level3/N으로표현.py
def solution(N, number):
    if N == number:
        return 1
    
    n_list = {}
    for i in range(1,9):
        n_list[i] = [int(str(N)*i)]
    
    answer = -1
    for i in range(2,9):
        for j in range(1,(i//2)+1):
            for x in n_list[j]:
                for y in n_list[i-j]:
                    n_list[i].append(x+y)
                    n_list[i].append(x-y)
                    n_list[i].append(y-x)
                    n_list[i].append(x*y)
                    if x != 0:
                        n_list[i] = n_list[i]+[y//x] if y/x == y//x else n_list[i]
                    if y != 0:
                        n_list[i] = n_list[i]+[x//y] if x/y == x//y else n_list[i]
        n_list[i] = list(set(n_list[i]))
        if number in n_list[i]:
            answer = i
            break
    return answer

File: level3/N으로표현.cpp
#include <string>
#include <vector>
#include <cmath>

using namespace std;

int makeNum(int N, int digit){
    int num = 0;
    for(int i = 0; i < digit; i++){
        num += pow(10,i)*N;
    }
    return num;
}

int solution(int N, int number) {
    if(N == number){
        return 1;
    }
    
    vector<vector<int>> v;
    for(int i = 0; i < 9; i++){
        vector<int> tmp;
        if(i == 0){
            v.push_back(tmp);
            continue;
        }
        tmp.push_back(makeNum(N,i));
        v.push_back(tmp);
    }
    
    int answer = -1;
    for(int i = 2; i <= 8; i++){
        for(int j = 1; j <= i/2; j++){
            for(int x = 0; x < v[j].size(); x++){
                for(int y = 0; y < v[i-j].size(); y++){
                    v[i].push_back(v[j][x]+v[i-j][y]);
                    v[i].push_back(v[j][x]-v[i-j][y]);
                    v[i].push_back(v[i-j][y]-v[j][x]);
                    v[i].push_back(v[j][x]*v[i-j][y]);
                    if(v[i-j][y] != 0){
                        v[i].push_back(v[j][x]/v[i-j][y]);
                    }
                    if(v[j][x] != 0){
                        v[i].push_back(v[i-j][y]/v[j][x]);
                    }
                }
            }
        }
        for(int j = 0; j < v[i].size(); j++){
            if(v[i][j] == number){
                answer = i;
                break;
            }
        }
        if(answer != -1){
            break;
        }
    }
    
    return answer;
}
